File: web/src/lib/growth.ts
// Tree establishment / growth model for the age slider.
//
// This MIRRORS coolspend/cost_model.py::growth_cooling_fraction exactly — a
// linear ramp from `initial_fraction` at planting to 1.0 at `ramp_years`:
//
//   frac(y) = initial + (1 - initial) * min(y / ramp_years, 1)
//
// It is the project's single, sourced growth curve (DECLARED linear ramp; see
// cost_model GrowthDiscountParams). We do NOT invent a second biological
// crown-diameter curve: the same fraction scales both the cooling readout and
// the visual canopy glyph, so the slider stays consistent with the cost model.

export interface GrowthParams {
  ramp_years: number
  initial_fraction: number
  horizon_years: number
}

export const DEFAULT_GROWTH: GrowthParams = {
  ramp_years: 25,
  initial_fraction: 0.2,
  horizon_years: 40,
}

/** Effective cooling fraction (0..1) at a given age in years. */
export function coolingFraction(year: number, p: GrowthParams = DEFAULT_GROWTH): number {
  const ramp = p.ramp_years > 0 ? Math.min(year / p.ramp_years, 1) : 1
  return p.initial_fraction + (1 - p.initial_fraction) * ramp
}

/**
 * Visual canopy scale (0..1) for the tree glyph at a given age. We scale the
 * glyph by the same establishment fraction (labelled as such in the UI) so the
 * canopy visibly grows with the slider without introducing an unsourced curve.
 */
export function canopyScale(year: number, p: GrowthParams = DEFAULT_GROWTH): number {
  return coolingFraction(year, p)
}

// ── Species-specific allometric growth (Chapman–Richards) ───────────────────
// Mirrors coolspend/growth.py exactly: crown(age) grows sigmoidally from a
// nursery planting crown to the species' mature crown, reaching ~95% at the
// species' maturity year. Used to grow each tree's canopy disk at its OWN pace
// (a fast Tipuana fills in ~2x sooner than a slow Cercis) instead of one flat
// global ramp. Anchored to real per-species values (mature crown + maturity
// band), not invented coefficients.

export const PLANTING_CROWN_M = 1.5 // large-caliper nursery stock crown at age 0
export const MAX_MATURITY_YEARS = 40 // slowest band — the slider's upper bound
const SHAPE_P = 3.0
const MATURITY_FRACTION = 0.95

function rateK(maturityYears: number): number {
  if (maturityYears <= 0) return 1.0
  const inner = 1 - Math.pow(MATURITY_FRACTION, 1 / SHAPE_P) // = e^(−k·T)
  return -Math.log(inner) / maturityYears
}

/** Mature-anchored crown diameter (m) at a given age after planting. */
export function crownDiameterAtAge(
  matureCrownM: number,
  maturityYears: number,
  ageYears: number,
): number {
  if (ageYears <= 0) return PLANTING_CROWN_M
  const k = rateK(maturityYears)
  const raw = matureCrownM * Math.pow(1 - Math.exp(-k * ageYears), SHAPE_P)
  return Math.max(PLANTING_CROWN_M, Math.min(matureCrownM, raw))
}

/** Fraction of mature shade delivered at a given age (crown AREA ratio, 0..1). */
export function coolingFractionAtAge(
  matureCrownM: number,
  maturityYears: number,
  ageYears: number,
): number {
  if (matureCrownM <= 0) return 1
  const r = crownDiameterAtAge(matureCrownM, maturityYears, ageYears) / matureCrownM
  return Math.min(1, r * r)
}

File: coolspend/growth.py
"""Species-specific canopy growth over time (the "years after planting" model).

PAPER limitation (growth accuracy): the old establishment curve was a single
flat *linear* ramp (0.2 → 1.0 over 25 yr) applied to every species. Real crown
growth is sigmoid and species-paced. This module replaces the linear ramp with
a **Chapman–Richards** growth function — the standard forestry/urban-tree growth
form — anchored to two values we already hold per species:

    * mature crown diameter  (coolspend.bcn_species, from the BCN inventory bands)
    * time-to-near-mature-canopy (coolspend.ecology.maturity_years: 20/30/40 yr)

We do NOT invent per-species fitted coefficients. We fit the single rate
parameter ``k`` so the curve reaches 95 % of the mature crown at the species'
maturity year, start from a realistic large-caliper nursery crown, and use the
classic sigmoid shape exponent. The result is a defensible parametric curve per
species — strictly more honest than one linear ramp — that can later be swapped
for measured i-Tree/Pretzsch species coefficients without changing the call
sites.

Form:  crown(age) = max(PLANTING_CROWN_M, A · (1 − e^(−k·age))^P)
where A = mature crown, P = 3 (sigmoid), k solved from maturity_years.
Cooling/shade scales with crown AREA, so cooling_fraction = (crown/A)^2.
"""
from __future__ import annotations

import math

# Large-caliper (20–25 cm) nursery stock — the planting crown at age 0. This is
# the standard street-tree planting stock the cost model already prices (€600
# tree_stock line). Crown of such stock is ~1.5 m.
PLANTING_CROWN_M: float = 1.5

# Sigmoid shape exponent of the Chapman–Richards form. P=3 gives the slow-fast-
# plateau shape typical of broadleaf urban trees.
_SHAPE_P: float = 3.0

# We anchor "near-mature" at 95 % of the mature crown at maturity_years.
_MATURITY_FRACTION: float = 0.95


def _rate_k(maturity_years: float) -> float:
    """Solve k so (1 − e^(−k·T))^P = 0.95 at T = maturity_years."""
    if maturity_years <= 0:
        return 1.0
    inner = 1.0 - _MATURITY_FRACTION ** (1.0 / _SHAPE_P)  # = e^(−k·T)
    return -math.log(inner) / maturity_years


def crown_diameter_at_age(scientific: str, age_years: float) -> float:
    """Mature-anchored crown diameter (m) at a given age after planting."""
    from coolspend.bcn_species import get_species  # noqa: PLC0415
    from coolspend.ecology import maturity_years  # noqa: PLC0415

    sp = get_species(scientific)
    mature = sp.crown_diameter_m if sp else 6.0
    if age_years <= 0:
        return PLANTING_CROWN_M
    k = _rate_k(float(maturity_years(scientific)))
    raw = mature * (1.0 - math.exp(-k * age_years)) ** _SHAPE_P
    return max(PLANTING_CROWN_M, min(mature, raw))


def cooling_fraction_at_age(scientific: str, age_years: float) -> float:
    """Fraction of mature shade/cooling delivered at a given age (0–1).

    Shade scales with crown projected AREA, so this is (crown(age)/mature)^2.
    At age 0 it is the (small) nursery-crown fraction, not zero.
    """
    from coolspend.bcn_species import get_species  # noqa: PLC0415

    sp = get_species(scientific)
    mature = sp.crown_diameter_m if sp else 6.0
    if mature <= 0:
        return 1.0
    return min(1.0, (crown_diameter_at_age(scientific, age_years) / mature) ** 2)


def growth_params(scientific: str) -> dict:
    """Per-species curve parameters, for the frontend to render the same curve.

    Returns the values the JS port (web/src/lib/growth.ts) needs to compute the
    identical Chapman–Richards crown(age) live for the age slider.
    """
    from coolspend.bcn_species import get_species  # noqa: PLC0415
    from coolspend.ecology import maturity_years  # noqa: PLC0415

    sp = get_species(scientific)
    mature = sp.crown_diameter_m if sp else 6.0
    my = float(maturity_years(scientific))
    return {
        "mature_crown_m": mature,
        "maturity_years": my,
        "planting_crown_m": PLANTING_CROWN_M,
        "shape_p": _SHAPE_P,
        "rate_k": round(_rate_k(my), 5),
    }
